src/nvim/eval.c

```diff
@@ -10199,22 +10199,149 @@ static void f_getcmdwintype(typval_T *argvars, typval_T *rettv)
+      return;
+    }
+    scope_number[i] = argvars[i].vval.v_number;
+    // The scope is the current iteration step.
+    scope = i;
+    // It is an error for the scope number to be less than `-1`.
+    if (scope_number[i] < -1) {
+      EMSG(_(e_invarg));
+      return;
+    }
+  }
+
+  // Allocate and initialize the string to return.
   cwd = xmalloc(MAXPATHL);
-  if (os_dirname(cwd, MAXPATHL) != FAIL) {
-    rettv->vval.v_string = vim_strsave(cwd);
+
+  // Get the scope and numbers from the arguments
+  for (int i = 0; i < MAX_CD_SCOPE; i++) {
+    // If there is no argument there are no more scopes after it, break out.
+    if (argvars[i].v_type == VAR_UNKNOWN) {
+      break;
+    }
+    scope_number[i] = argvars[i].vval.v_number;
+    // The scope is the current iteration step.
+    scope = i;
+    // It is an error for the scope number to be less than `-1`.
+    if (scope_number[i] < -1) {
+      EMSG(_(e_invarg));
+      goto fail;
+    }
+  }
+
+  // It the highest scope number is `-1` lower the scope by one.
+  if (scope_number[scope] < 0) {
+    scope++;
+  }
+
+  // Find the tabpage by number
+  if (scope_number[kCdScopeTab] == -1) {
+    tp = NULL;
+  } else if (scope_number[kCdScopeTab] > 0) {
+    tp = find_tabpage(scope_number[kCdScopeTab]);
+    if (!tp) {
+      EMSG(_("E740: Cannot find tab number."));
+      goto fail;
+    }
+  }
+
+  // Find the window in `tp` by number, `NULL` if none.
+  if (scope_number[kCdScopeWindow] == -1) {
+    win = NULL;
+  } else if (scope_number[kCdScopeWindow] >= 0) {
+    // If `tp` is `NULL` that's an error
+    if (!tp) {
+      EMSG(_("E740: Second argument cannot be -1 if first argument is >= 0."));
+      goto fail;
+    }
+
+    if (scope_number[kCdScopeWindow] > 0) {
+      win = find_win_by_nr(&argvars[0], curtab);
+      if (!win) {
+        EMSG(_("E740: Cannot find window number."));
+        goto fail;
+      }
+    }
+  }
+
+  // Act according to the scope. Add new scopes to the switch as well.
+  switch (scope) {
+    case kCdScopeWindow:
+      from = win->w_localdir;
+      if (from) {
+        break;
+      }
+      // Fall through intentionally
+    case kCdScopeTab:
+      from = tp->localdir;
+      if (from) {
+        break;
+      }
+      // Fall through intentionally
+    case kCdScopeGlobal:
+      // The `globaldir` variable is not always set.
+      if (globaldir) {
+        from = globaldir;
+      } else {
+        // Copy the OS path dircetly into output string and jump to the end.
+        if (os_dirname(cwd, MAXPATHL) == FAIL) {
+          EMSG(_("E41: Could not display path."));
+          goto fail;
+        }
+      }
+      break;
+  }
+
+  // Copy the working directory string to `cwd` for print
+  if (from) {
+    xstrlcpy((char *)cwd, (char *)from, MAXPATHL);
+  }
+
+  rettv->vval.v_string = vim_strsave(cwd);
 #ifdef BACKSLASH_IN_FILENAME
-    slash_adjust(rettv->vval.v_string);
+  slash_adjust(rettv->vval.v_string);
 #endif
-  }
+
+fail:
```

Comment:
`end` would be a better name for this label, because it falls through to this in the non-failure case. Most of the existing code is consistent about this AFAICT.